tutorial/decorators.md

```diff
@@ -0,0 +1,243 @@
+
+This decorator indicates that all arrays mentioned as argements (of the decorator) should be stored
+on the stack.
+
+This example shows how the decorators can affect the conversion of the array between the supported languages, Pyccel here is told by the decorator `stack array` to store the array `arrat_in_stack` in the stack, For the array `array_in_heap` Pyccel is assuming that it should be stored in the heap:
+
+```python
+from pyccel.decorators import stack_array
+
+@stack_array('array_in_stack')
+def fun1():
+
+     #/////////////////////////
+     #array stored in the stack
+     #////////////////////////
+     array_in_stack = [1,2,3]
+     #////////////////////////
+     #array stored in the heap
+     #////////////////////////
+     array_in_heap = [1,2,3]
+```
+
+This the C generated code:
+
+```C
+
+#include "boo.h"
+#include <stdint.h>
+#include <stdlib.h>
+#include "ndarrays.h"
+#include <string.h>
+
+
+/*........................................*/
+void fun1(void)
+{
+    int64_t array_dummy_0003[3];
+    t_ndarray array_in_stack = (t_ndarray){
+        .nd_int64=array_dummy_0003,
+        .shape=(int64_t[]){3},
+        .strides=(int64_t[1]){0},
+        .nd=1,
+        .type=nd_int64,
+        .is_view=false
+    };
+    stack_array_init(&array_in_stack);
+    t_ndarray array_in_heap;
+    /*/////////////////////////*/
+    /*array stored in the stack*/
+    /*////////////////////////*/
+    int64_t array_dummy_0001[] = {1, 2, 3};
+    memcpy(array_in_stack.nd_int64, array_dummy_0001, array_in_stack.buffer_size);
+    /*////////////////////////*/
+    /*array stored in the heap*/
+    /*////////////////////////*/
+    array_in_heap = array_create(1, (int64_t[]){3}, nd_int64);
+    int64_t array_dummy_0002[] = {1, 2, 3};
+    memcpy(array_in_heap.nd_int64, array_dummy_0002, array_in_heap.buffer_size);
+}
+/*........................................*/
+```
+
+This the Fortran equivalent code:
+
+```Fortran
+module boo
+
+
+  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+  implicit none
+
+  contains
+
+  !........................................
+  subroutine fun1()
+
+    implicit none
+
+    integer(i64) :: array_in_stack(0:2_i64)
+    integer(i64), allocatable :: array_in_heap(:)
+
+    !/////////////////////////
+    !array stored in the stack
+    !////////////////////////
+    array_in_stack = [1_i64, 2_i64, 3_i64]
+    !////////////////////////
+    !array stored in the heap
+    !////////////////////////
+    allocate(array_in_heap(0:2_i64))
+    array_in_heap = [1_i64, 2_i64, 3_i64]
+
+  end subroutine fun1
+  !........................................
+
+end module boo
+```
+
+## Allow negative index
+
+This one indicates that all arrays mentioned as argements (of the decorator) can be accessed with negative indexes. unlike fortran you can't use negative indexes in arrays in C, But with Pyccel you can work comfotabally with them in the Python code, For the generated C code Pyccel will take care of that.
```

Comment:
Typos: `argements`->`arguments`   `comfotabally`->`comfortably `

Fortran allows negative indexes but they do not play the same role as they do in python. In fortran you can choose what number is the first index. For example:
```
allocate(my_array(-5:-3))
my_array = [1_i64, 2_i64, 3_i64]
print *, my_array(-5)
```
will print `1`. Negative indexing is therefore only allowed if the array was defined to have negative indexes.
When translating python code to fortran all our arrays have `0` as the first index therefore we cannot use negative indexes even in Fortran.

I don't think this description is very clear. I suggest:

```
In python negative indexes allow a user to index an array starting from the back (e.g. the index -1 is the
last element of the array). Pyccel recreates this behaviour for literal indexes. However when an array is
indexed with a variable or an expression, it is impossible to know at compile time whether the index is
positive or negative. As a result an if block must be added. This implies a (potentially large) performance
cost. Non-literal negative indexes are not especially common, therefore Pyccel does not add this costly
if block unless it is specifically requested. This can be done using the `allow_negative_index` decorator
```